app/warden/magic_link_strategy.rb

```diff
@@ -1,15 +1,28 @@
 class MagicLinkStrategy < Warden::Strategies::Base
+  MAX_TOKEN_DURATION = 24.hours
+
   def valid?
     params[:token].present?
   end
 
   def authenticate!
     candidate = FindCandidateByToken.call(raw_token: params[:token])
 
-    if candidate
+    if candidate.present? && token_not_expired?(candidate)
+      expire_token(candidate)
       success!(candidate)
     else
       fail!
     end
   end
+
+private
+
+  def expire_token(candidate)
+    candidate.update!(magic_link_token: nil, magic_link_token_sent_at: nil)
```

Comment:
This feels close to the database, and therefore there's an argument that it should be on the model.

I don't like the idea of defining it in `Candidate`, though as that breaks the boundary between `MagicLink` and `Candidate`.

I think the correctest way of doing it might be to define `:magic_link` as a proper devise module and mix it in the conventional way. If that's painful and/or overkill, would it be excessive to introduce a suitable-namespaced concern for these two methods and mix it into `Candidate`?

Reply:
I wonder if it's okay in this class - it didn't surprise me when reading it initially. It basically tacks on something to the `success!` method, doing it in another way might be harder to follow.

Reply:
This sounds good! I'm not very familiar with devise modules, could you please reply with some brief pseudo-code with what files/concerns we'd have to create and what goes where to help me along?

Reply:
Update: this is actually being removed due to deciding not to expire tokens in this way anymore due to prefetching/Slack preview et al.